The Inversenning : Superior Healing Medications

MGOOOOOO opened the pull request:
## About The Pull Request

Please read the contents of my HackMD to see what my PR does, as well as further elaboration on why it's good for the game.

https://hackmd.io/@MGO/r1idU4cza

## Contents
Introducing new inverse reagents for existing superior healing medications! This push includes...

**Benzoic Acid** : Inverse of Salicylic Acid. Robust fertilizer that provides a decent range of benefits for plant life.
![ss13-1](https://github.com/tgstation/tgstation/assets/97645027/6d624397-f860-47f2-9702-a073c2a936c2)

**Oxymetholone** : Inverse of Oxandrolone. Anabolic steroid that promotes the growth of muscle during and after exercise.
![ss13-2](https://github.com/tgstation/tgstation/assets/97645027/4a4c941a-e5d1-4aed-9785-9c0f702ede2a)

**Bamethan** : Inverse of Salbutamol. Blood thinner that drastically increases the chance of receiving bleeding wounds.
![ss13-3](https://github.com/tgstation/tgstation/assets/97645027/1ae9c6be-a181-42e6-b2e4-45e7dbee694e)

**Pendetide** : Inverse of Pentetic Acid. An unusual bioradioactive drug that purges basic radiation healing chems. Also increases the severity of radiation poisoning.
![ss13-4](https://github.com/tgstation/tgstation/assets/97645027/6ad81590-69bb-4f09-b166-1e34c5db168a)

**Hyoscyamine** : Inverse of Atropine. Heals heart and stomach damage, and slowly removes minor toxin damage.
![ss13-7](https://github.com/tgstation/tgstation/assets/97645027/5823d8dd-11d6-4225-96e6-2ace673474a4)

**Ammoniated Sludge** : Inverse of Ammoniated Mercury. A ghastly looking mess of mercury by-product which causes bursts of manic hysteria.
![ss13-6](https://github.com/tgstation/tgstation/assets/97645027/a1966766-673d-49cc-a3f0-f103d733c7b5)

**Inreziniver** : Inverse of Rezadone. Makes the user horribly afraid of all things related to carps.
![ss13-8](https://github.com/tgstation/tgstation/assets/97645027/1e6f926b-50d3-47c7-bdde-60cb5da6409a)

## Why It's Good For The Game

This is an effort to add more variety to the existing inverse reagent system within chemistry. Not only should this variety serve to provide additional options for a Chemist to experiment with, they should also broaden the possibilities for already existing strategies.

## Changelog
:cl:MGO
add: Introduces new inverse reagents for Salicylic Acid, Oxandrolone, Salbutamol, Pentetic Acid, Atropine, Ammoniated Mercury and Rezadone!
/:cl:

MGOOOOOO committed:
The Inversenning : Superior Healing Medications

MGOOOOOO committed:
Merge branch 'master' into MGOChooms2023

trerri commented:
I think coders are trying to move away from cellular damage and replacing it with other effects. For example slimes and the slot machine of greed have been changed to not use cellular damage.

MGOOOOOO commented:
> I think coders are trying to move away from cellular damage and replacing it with other effects. For example slimes and the slot machine of greed have been changed to not use cellular damage.

Ah yeah, I just found the PR that removed most forms of cellular damage. (https://github.com/tgstation/tgstation/pull/77832) I'll go ahead and replace it with a more suitable effect, one moment...

MGOOOOOO commented:
Okay, as far as Inreziniver goes, I can think of two suitable replacements:
1) A powerful neuron fortifying medication that prevents brain death within the user. Will not heal the brain and absorption rate is very fast. (This was my old version of Inreziniver. Might be too strong...)

2) Provides a unique phobia of space carp and all carp related items, like clothing and plushies.

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+	ph = 3.4
+	tox_damage = 0
+
+/datum/reagent/inverse/sal_acid/on_hydroponics_apply(obj/item/seeds/myseed, datum/reagents/chems, obj/machinery/hydroponics/mytray)
+	. = ..()
+	if(myseed && chems.has_reagent(src.type, 1))
+		myseed.adjust_potency(round(chems.get_reagent_amount(src.type) * 0.25))
+		myseed.adjust_yield(round(chems.get_reagent_amount(src.type) * 0.2))
+		myseed.adjust_production(-round(chems.get_reagent_amount(src.type) * 0.2))
+		mytray.adjust_plant_health(round(chems.get_reagent_amount(src.type) * 0.5))
```
I'm not sure if this code is old or you're upstreaming this from an older codebase but this is the outdated form of `on_hydroponics_apply` and won't work. 

This is how the proc is structured now. 

![image](https://github.com/tgstation/tgstation/assets/51863163/84bf1959-1eaf-4385-aaf2-2d474b73d898)

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+		affected_mob.adjustStaminaLoss(5 * REM * seconds_per_tick)
+	else if(SPT_PROB(5, seconds_per_tick))
+		affected_mob.vomit(VOMIT_CATEGORY_BLOOD, lost_nutrition = 0, distance = 3)
+		//affected_mob.adjustBruteLoss(10 * REM * seconds_per_tick, updating_health = FALSE)
```
No commented out code please.

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+
+/datum/reagent/inverse/salbutamol/on_mob_metabolize(mob/living/L)
+	. = ..()
+	ADD_TRAIT(L,TRAIT_BAMETHAN_BLEED,type)
```
```suggestion
	ADD_TRAIT(L, TRAIT_BAMETHAN_BLEED, type)
```

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+
+/datum/reagent/inverse/salbutamol/on_mob_end_metabolize(mob/living/M)
+	. = ..()
+	REMOVE_TRAIT(M,TRAIT_BAMETHAN_BLEED,type)
```
```suggestion
	REMOVE_TRAIT(M, TRAIT_BAMETHAN_BLEED, type)
```

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+	tox_damage = 0
+
+/datum/reagent/inverse/pen_acid/on_mob_life(mob/living/carbon/affected_mob, seconds_per_tick, times_fired)
+	if(holder.has_reagent(/datum/reagent/medicine/c2/seiver))
+		holder.remove_reagent(/datum/reagent/medicine/c2/seiver, 5 * REM * seconds_per_tick)
+	if(holder.has_reagent(/datum/reagent/medicine/potass_iodide))
+		holder.remove_reagent(/datum/reagent/medicine/potass_iodide, 5 * REM * seconds_per_tick)
+	if(holder.has_reagent(/datum/reagent/medicine/c2/multiver))
+		holder.remove_reagent(/datum/reagent/medicine/c2/multiver, 5 * REM * seconds_per_tick)
```
It should be safe to call these without them actually being in the holder's system.
```suggestion
	holder.remove_reagent(/datum/reagent/medicine/c2/seiver, 5 * REM * seconds_per_tick)
	holder.remove_reagent(/datum/reagent/medicine/potass_iodide, 5 * REM * seconds_per_tick)
	holder.remove_reagent(/datum/reagent/medicine/c2/multiver, 5 * REM * seconds_per_tick)
```

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+		affected_mob.adjust_disgust(3 * REM * seconds_per_tick)
+		if(SPT_PROB(2.5, seconds_per_tick))
+			to_chat(affected_mob, span_warning("A horrible ache spreads in your insides!"))
+			affected_mob.adjust_confusion(10 SECONDS)
```
Might want to cap this confusion otherwise it has some RNG potential of being incredibly obnoxious

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+	var/need_mob_update
+	need_mob_update = affected_mob.adjustOrganLoss(ORGAN_SLOT_STOMACH, -1 * REM * seconds_per_tick)
+	need_mob_update += affected_mob.adjustOrganLoss(ORGAN_SLOT_HEART, -1 * REM * seconds_per_tick)
+	if(need_mob_update)
+		. = UPDATE_MOB_HEALTH
+	if(affected_mob.getToxLoss() <= 25)
+		need_mob_update = affected_mob.adjustToxLoss(-0.5, updating_health = FALSE, required_biotype = affected_biotype)
```
You're setting `need_mob_update` after you check for it
```suggestion
	if(affected_mob.getToxLoss() <= 25)
		need_mob_update = affected_mob.adjustToxLoss(-0.5, updating_health = FALSE, required_biotype = affected_biotype)
	if(need_mob_update)
		return UPDATE_MOB_HEALTH
```

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+	. = ..()
+	if(SPT_PROB(7.5, seconds_per_tick))
+		affected_mob.emote("scream")
+		affected_mob.say(pick("AAAAAAAHHHHH!!","OOOOH NOOOOOO!!","GGGUUUUHHHHH!!","AIIIIIEEEEEE!!","HAHAHAHAHAAAAAA!!","OORRRGGGHHH!!","AAAAAAAJJJJJJJJJ!!"), forced = /datum/reagent/inverse/ammoniated_mercury)
```
```suggestion
		affected_mob.say(pick("AAAAAAAHHHHH!!","OOOOH NOOOOOO!!","GGGUUUUHHHHH!!","AIIIIIEEEEEE!!","HAHAHAHAHAAAAAA!!","OORRRGGGHHH!!","AAAAAAAJJJJJJJJJ!!"), forced = type)
```

MrMelbert commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,148 @@ Basically, we fill the time between now and 2s from now with hands based off the
+/datum/reagent/inverse/rezadone/on_mob_life(mob/living/carbon/affected_mob, seconds_per_tick, times_fired)
+	. = ..()
+	if(affected_mob.getToxLoss() >= 1.5)
+		var/need_mob_update
+		need_mob_update = affected_mob.adjustToxLoss(-1.5 * REM * seconds_per_tick, updating_health = FALSE)
+		need_mob_update += affected_mob.adjustCloneLoss((0.25 + creation_purity) * REM * seconds_per_tick, updating_health = FALSE)
```
You're tracking updates but not returning `UPDATE_MOB_HEALTH` if any happened
```suggestion
		var/need_mob_update
		need_mob_update = affected_mob.adjustToxLoss(-1.5 * REM * seconds_per_tick, updating_health = FALSE)
		need_mob_update += affected_mob.adjustCloneLoss((0.25 + creation_purity) * REM * seconds_per_tick, updating_health = FALSE)
		if(need_mob_update)
			return UPDATE_MOB_HEALTH
```

MrMelbert commented on code/modules/surgery/bodyparts/wounds.dm:
```diff
@@ -67,6 +67,10 @@
 	if(HAS_TRAIT(owner, TRAIT_EASYDISMEMBER))
 		damage *= 1.1
 
+	//For inverse reagent Bamethan!
+	if(HAS_TRAIT(owner, TRAIT_BAMETHAN_BLEED) && ((woundtype == WOUND_PIERCE) || (woundtype == WOUND_SLASH)))
+		damage *= 1.5
```
Consider making this a generic trait so other people can use it: `TRAIT_EASILY_WOUNDED_ONLY_BLEED_WOUNDS`, as a verbose example. 

Not only does it make it re-usable but also easier to understand what it is doing at a glance from the other side.

MrMelbert commented on code/_globalvars/traits.dm:
```diff
@@ -76,6 +76,7 @@ GLOBAL_LIST_INIT(traits_by_type, list(
 		"TRAIT_EASILY_WOUNDED" = TRAIT_EASILY_WOUNDED,
 		"TRAIT_HARDLY_WOUNDED" = TRAIT_HARDLY_WOUNDED,
 		"TRAIT_NEVER_WOUNDED" = TRAIT_NEVER_WOUNDED,
+		"TRAIT_BAMETHAN_BLEED" = TRAIT_BAMETHAN_BLEED, // Bleed modifier trait for Bamethan
```
You don't need a doc comment in this list

MrMelbert commented on code/__DEFINES/traits.dm:
```diff
@@ -302,6 +302,7 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_EASILY_WOUNDED "easy_limb_wound"
 #define TRAIT_HARDLY_WOUNDED "hard_limb_wound"
 #define TRAIT_NEVER_WOUNDED "never_wounded"
+#define TRAIT_BAMETHAN_BLEED "bamethan_bleed" // Bleed modifier trait for Bamethan
```
Make this an auto-doc comment

HalfPunch commented:
I like the idea of adding plant related chems into inverse reaction products. But, i think that there is no reason in adding chemical which is just diethylamine+saltpetre. It has no special effects and i think that it will not be used in rounds ever, at least when people will know what it does. But that is just my opinion.
What i really want to say is that we don't need strong botany chems, we need special, unique botany chem, ones that do some really interesting stuff.

MGOOOOOO commented:
> I like the idea of adding plant related chems into inverse reaction products. But, i think that there is no reason in adding chemical which is just diethylamine+saltpetre. It has no special effects and i think that it will not be used in rounds ever, at least when people will know what it does. But that is just my opinion. What i really want to say is that we don't need strong botany chems, we need special, unique botany chem, ones that do some really interesting stuff.

I'd like to continue making new inverse reagents in the future, and more unique botanical effects would be a great idea for other chemicals. I decided to make Benzoic the "baseline" inverse botanical product just to provide another option for fertilizing plants in general, though if you have any additional ideas I'd love to see them.

MGOOOOOO committed:
Updates code, reworks Inreziniver, slightly nerfs Pendetide

MGOOOOOO committed:
Merge branch 'MGOChooms2023' of https://github.com/MGOOOOOO/tgstation into MGOChooms2023
update

MGOOOOOO commented:
Man, I hope I did these additional commits right...

Thank you for the review MrMelbert, I implemented the changes you requested including the nerf to Pendetide's confusion as well. I also reworked Inreziniver entirely since cellular damage is being phased out, I have updated the HackMD and will post it here as well:

Inreziniver
Normal : Rezadone
Description : Makes the user horribly afraid of all things related to carps.
Consumption Rate : 0.05u/tick
Inverse Chem % : 25%
pH : 13.9

Assigns a phobia of carps to the consumer that can only be removed once the reagent is used up.

MGOOOOOO committed:
Merge branch 'master' into MGOChooms2023

HalfPunch commented:
> though if you have any additional ideas I'd love to see them

I'd like to see chem that will reduce plants endurance and lifespan just enough to make inverse chems inside plants. So it will be much more common to see inverse chem shenanigans from botanists. 
And since we will get inverse Atropine in this PR, it will be so cool if we could get Hyoscyamine from low endurance/lifespan Omega weed.

MGOOOOOO commented:
Hell, that'd be interesting to see. Botanists can already make some crazy plants for the station, why not toss some inverse reagents into the mix as well? Coveroli weed for impromptu atmos protection, Cyrogelidia apples for instant freeze blocks, and so on... I'd like to do that in the future as well!

MGOOOOOO committed:
Merge branch 'master' into MGOChooms2023

MGOOOOOO commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,143 @@ Basically, we fill the time between now and 2s from now with hands based off the
+
+/datum/reagent/inverse/atropine/overdose_process(mob/living/carbon/affected_mob, seconds_per_tick, times_fired)
+	. = ..()
+	affected_mob.adjustOrganLoss(pick(ORGAN_SLOT_HEART,ORGAN_SLOT_LIVER,ORGAN_SLOT_LUNGS,ORGAN_SLOT_STOMACH,ORGAN_SLOT_EYES,ORGAN_SLOT_EARS,ORGAN_SLOT_BRAIN,ORGAN_SLOT_APPENDIX,ORGAN_SLOT_TONGUE),2 * seconds_per_tick)
```
You're definitely right, I'll try to make these sort of long lines cleaner to read in the future, thanks!

MGOOOOOO committed:
Merges suggestions from JohnFulpWillard

MGOOOOOO committed:
Merge branch 'MGOChooms2023' of https://github.com/MGOOOOOO/tgstation into MGOChooms2023

MGOOOOOO committed:
Merge branch 'master' into MGOChooms2023

JohnFulpWillard commented on code/modules/reagents/chemistry/reagents/impure_reagents/impure_medicine_reagents.dm:
```diff
@@ -790,3 +790,154 @@ Basically, we fill the time between now and 2s from now with hands based off the
+		ORGAN_SLOT_APPENDIX,
+		ORGAN_SLOT_TONGUE,
+	)
+	affected_mob.adjustOrganLoss(pick(possible_organs) ,2 * seconds_per_tick)
```
Doesn't this revive non functional organs?

MGOOOOOO replied:
This function doesn't heal organs, I've tried it again just to be sure and it damages the organs correctly.

MGOOOOOO committed:
The epic

MGOOOOOO commented on code/_globalvars/phobias.dm:
```diff
@@ -234,6 +239,29 @@ GLOBAL_LIST_INIT(phobia_objs, list(
+		/obj/item/clothing/mask/gas/carp,
+		/obj/item/clothing/mask/cigarette/carp,
+		/obj/item/clothing/under/suit/carpskin,
+		/obj/item/dna_probe/carp_scanner,
```
No worries, I went ahead and removed the carp scanner from the list.

MGOOOOOO commented on code/_globalvars/phobias.dm:
```diff
@@ -234,6 +239,29 @@ GLOBAL_LIST_INIT(phobia_objs, list(
 		/obj/machinery/iv_drip,
 	)),
+	"carps" = typecacheof(list(
+		/obj/item/book/granter/martial/carp,
```
True...

MGOOOOOO committed:
Merge branch 'master' into MGOChooms2023

MGOOOOOO committed:
uh

MGOOOOOO committed:
Merge branch 'MGOChooms2023' of https://github.com/MGOOOOOO/tgstation into MGOChooms2023

MGOOOOOO committed:
please god
